url47af6b1535_rolandotr_db_comparison_tgz-pJ8-methodology_HistoryJ8/src/utils/SpiderChart.java

```diff
@@ -0,0 +1,273 @@
+				new CryptoCalls(new IntegerRelation(), new NoScale<Integer>()),
+				new FinalSlowPhase(new FinalSlowPhaseRelation(), new NoScale<Boolean>()),
+		};
+
+		printSpiderChart(firstSet, attributes, "swiss_ski.tex");
+	}
+
+	public static void main3(String[] args) throws IOException {
+		DBProtocol best = new BestProtocol();
+		best.setNumberOfRounds(16);
+		DBProtocol tree = new TreeBasedProtocol(8);
+		tree.setNumberOfRounds(16);
+		DBProtocol hk = new MunillaAndPeinadoProtocol(0.75);
+		hk.setNumberOfRounds(16);
+		DBProtocol bc = new BrandsAndChaumProtocol();
+		bc.setNumberOfRounds(16);
+		DBProtocol[] firstSet = new DBProtocol[]{
+				best,tree,hk, bc,
+		};
+		Attribute[] attributes = new Attribute[]{
+				new MafiaFraudProbability(new ProbabilityRelation(), new LogScale(2)),
+				new DistanceFraudProbability(new ProbabilityRelation(), new LogScale(2)),
+				new TerroristFraudProbability(new ProbabilityRelation(), new LogScale(2)),
+				new Memory(new MemoryRelation(), new KbitsScale()),
+				new CryptoCalls(new IntegerRelation(), new NoScale<Integer>()),
+				new FinalSlowPhase(new FinalSlowPhaseRelation(), new NoScale<Boolean>()),
+		};
+
+		printSpiderChart(firstSet, attributes, "tree_hk_bc.tex");
+	}
+
+	public static void main4(String[] args) throws IOException {
+		DBProtocol best = new BestProtocol();
+		best.setNumberOfRounds(32);
+		DBProtocol poulidor = new PoulidorProtocol();
+		poulidor.setNumberOfRounds(32);
+		DBProtocol[] firstSet = new DBProtocol[]{
+				best,poulidor,
+		};
+		Attribute[] attributes = new Attribute[]{
+				new MafiaFraudProbability(new ProbabilityRelation(), new LogScale(2)),
+				new DistanceFraudProbability(new ProbabilityRelation(), new LogScale(2)),
+				new Memory(new MemoryRelation(), new KbitsScale()),
+		};
+
+		printSpiderChart(firstSet, attributes, "uniform_against_others.tex");
+	}
+
+	public static void main2(String[] args) throws IOException {
+		DBProtocol best = new BestProtocol();
+		best.setNumberOfRounds(128);
+		DBProtocol tree = new TreeBasedProtocol(7, DBProtocol.SIZE_OF_NONCES);
+		tree.setNumberOfRounds(128);
+		DBProtocol poulidor = new PoulidorProtocol();
+		poulidor.setNumberOfRounds(128);
+		DBProtocol tma = new TMAProtocol();
+		tma.setNumberOfRounds(128);
+		DBProtocol[] firstSet = new DBProtocol[]{
+				best,tree,poulidor,tma
+		};
+		Attribute[] attributes = new Attribute[]{
+				new MafiaFraudProbability(new ProbabilityRelation(), new LogScale(2)),
+				new DistanceFraudProbability(new ProbabilityRelation(), new LogScale(2)),
+				new TerroristFraudProbability(new ProbabilityRelation(), new LogScale(2)),
+				new Memory(new MemoryRelation(), new KbitsScale()),
+		};
+
+		printSpiderChart(firstSet, attributes, "tree_poulidor_tma.tex");
+	}
+	
+	public static void printSpiderChart(DBProtocol[] protocols, Attribute[] attributes, 
+			String fileName) throws IOException{
+		FileWriter writer = new FileWriter(fileName);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code contains a resource that might not have closed properly. A resource leak can slow down or crash your system.

**Fix**
Consider closing the following resource in a try-finally block: *writer*. The resource is referenced at lines 131, 132, 133, 134, 137, 139. The resource is closed at line 140. There are other execution paths that do not contain closure statements, for example, when SpiderChart.addProtocol throws an exception.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).